dockstore-webservice/src/main/java/io/dockstore/webservice/resources/EntryResource.java

```diff
@@ -438,79 +510,66 @@ public Entry setDiscourseTopic(@ApiParam(hidden = true) @Parameter(hidden = true
+     * size is {@link #PROCESSOR_PAGE_SIZE}
+     * @param loader
+     * @param processor
+     * @param allVersions
+     * @return the number of entries processed
+     */
+    private <T extends Entry> int loadAndProcessEntries(Function<Integer, List<T>> loader,
+        BiFunction<List<T>, Boolean, Void> processor, Boolean allVersions) {
+        final ProcessorProgress progress = new ProcessorProgress();
         while (!progress.done) {
             final TransactionHelper transactionHelper = new TransactionHelper(sessionFactory);
             transactionHelper.transaction(() -> {
-                final List<Tool> tools =
-                    this.toolDAO.findAllTools(progress.offset, LANGUAGE_VERSION_PROCESSOR_PAGE_SIZE);
-                if (tools.size() < LANGUAGE_VERSION_PROCESSOR_PAGE_SIZE) {
+                final List<T> list = loader.apply(progress.offset);
+                if (list.size() < PROCESSOR_PAGE_SIZE) {
                     progress.done = true;
                 }
-                progress.processedEntries += tools.size();
-                LOG.info("Executing {} tool language version updates starting at offset {}",
-                    LANGUAGE_VERSION_PROCESSOR_PAGE_SIZE, progress.offset);
-                progress.offset += LANGUAGE_VERSION_PROCESSOR_PAGE_SIZE;
-                tools.stream()
-                    .filter(tool -> tool.getDescriptorType().size() == 1) // Only support tools with 1 language
-                    .forEach(tool -> {
-                        final String descriptorLanguageText = tool.getDescriptorType().get(0);
-                        LanguageHandlerInterface languageHandlerInterface = LanguageHandlerFactory.getInterface(
-                            DescriptorLanguage.convertShortStringToEnum(descriptorLanguageText));
-                        tool.getWorkflowVersions().stream()
-                            .filter(version -> allVersions || version.getVersionMetadata().getDescriptorTypeVersions().isEmpty())
-                            .forEach(version -> {
-                                final String primaryPath;
-                                if (DescriptorLanguage.convertShortStringToEnum(descriptorLanguageText) == DescriptorLanguage.WDL) {
-                                    primaryPath = version.getWdlPath();
-                                } else { // We will not add new languages to Tool (perhaps to AppTool), so this is safe
-                                    primaryPath = version.getCwlPath();
-                                }
-                                readSourceFilesAndUpdate(languageHandlerInterface, version, primaryPath);
-                            });
-                    });
+                progress.processedEntries += list.size();
+                LOG.info("Executing {} updates starting at offset {}", list.size(), progress.offset);
+                progress.offset += PROCESSOR_PAGE_SIZE;
+                processor.apply(list, allVersions);
```

Comment:
Is the desired behavior to abort [and skip the processing of the remaining entries] when an exception is thrown here?

Reply:
I was thinking there would be no exceptions thrown, but there's one been one case where there has been an exception, so good point, surrounded with try/catch.